onnxruntime/core/providers/webnn/builders/impl/slice_op_builder.cc

```diff
@@ -0,0 +1,161 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Copyright (c) Intel Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/common/safeint.h"
+#include "core/framework/tensorprotoutils.h"
+#include "core/optimizer/initializer.h"
+#include "core/providers/common.h"
+#include "core/providers/cpu/tensor/slice_helper.h"
+#include "core/providers/shared/utils/utils.h"
+#include "core/providers/webnn/builders/helper.h"
+#include "core/providers/webnn/builders/model_builder.h"
+#include "core/providers/webnn/builders/op_builder_factory.h"
+
+#include "base_op_builder.h"
+
+namespace onnxruntime {
+namespace webnn {
+
+class SliceOpBuilder : public BaseOpBuilder {
+  // Add operator related.
+ public:
+  void AddInitializersToSkip(ModelBuilder& model_builder, const Node& node) const override;
+
+ private:
+  Status AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
+                               const logging::Logger& logger) const override ORT_MUST_USE_RESULT;
+  bool IsOpSupportedImpl(const InitializedTensorSet& initializers, const Node& node,
+                         const logging::Logger& logger) const override;
+  // TODO: Support Slice opset < 10, which uses attributes for starts and ends.
+  int GetMinSupportedOpSet(const Node& /* node */) const override { return 10; }
+};
+
+// Add operator related.
+
+void SliceOpBuilder::AddInitializersToSkip(ModelBuilder& model_builder, const Node& node) const {
+  // Skip all initializer except the first inputs(data).
+  for (size_t i = 1; i < node.InputDefs().size(); i++) {
+    model_builder.AddInitializerToSkip(node.InputDefs()[i]->Name());
+  }
+}
+
+Status SliceOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
+                                             const Node& node,
+                                             const logging::Logger& logger) const {
+  const auto& input_defs = node.InputDefs();
+  std::vector<int64_t> input_shape;
+  ORT_RETURN_IF_NOT(GetShape(*input_defs[0], input_shape, logger), "Cannot get shape");
+  auto rank = input_shape.size();
+  NodeAttrHelper helper(node);
+
+  emscripten::val inputs = model_builder.GetOperand(input_defs[0]->Name());
+  std::vector<int32_t> starts(rank, 0);
+  std::vector<int32_t> sizes(input_shape.begin(), input_shape.end());
+
+  // Copy the data from the starts/ends/axes/steps initializers.
+  TensorShapeVector input_starts;
+  TensorShapeVector input_ends;
+  TensorShapeVector input_axes;
+  TensorShapeVector input_steps;
+  SliceOp::PrepareForComputeMetadata compute_metadata(input_shape);
+  const auto CopyInputData = [&input_defs, &model_builder](size_t input_idx, TensorShapeVector& data) {
+    data.clear();
+
+    // This is an optional input, return empty vector.
+    if (input_defs.size() <= input_idx)
```

Comment:
Optional tensors can be indicated by an empty name too. e.g.:

```
    if (input_idx >= input_defs.size())
        return Status::OK();

    const auto& input_name = input_defs[input_idx]->Name();
    if (input_name.empty())
        return Status::OK();
```

Note `starts` and `ends` are not optional. So consider passing a boolean is-required parameter to the helper function . e.g. `ORT_RETURN_IF_ERROR(CopyInputData(1, input_starts, true));` vs `  ORT_RETURN_IF_ERROR(CopyInputData(3, input_axes, false));`

Btw, Bruce Dai Feng had a similar related function for reading tensors into a 1D array for Split. Maybe you can mostly reuse that here?

Reply:
> Btw, Bruce Dai Feng had a similar related function for reading tensors into a 1D array for Split. Maybe you can mostly reuse that here?

Umm, there's a bit difference that `Slice` accepts both `Int32` and `Int64` tensor, I will try to make it compatible.

Reply:
No big deal if it's not used - just noticing some similar patterns and wanting to reduce code duplication.

We also have a little helper in the DML EP that might be useful for reference (though it reads into int32 instead of uint32): https://github.com/microsoft/onnxruntime/blob/5abaca9d6969f65998708abf6cbc7616036b1769/onnxruntime/core/providers/dml/OperatorAuthorHelper/OperatorHelper.cpp#L132-L174